### ISSUE TITLE
Buggy regexp generation when using optionals

### ISSUE THREAD

**Issue by the3coder** [aichaos/rivescript-python#18]
When requesting a reply from the rivescript script, when using nested optionals a buggy regular expression is created by rivescript, which causes the re library to throw an exception.

The python traceback prints the following:

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/easyirc/event/__init__.py", line 14, in putln
    consumed = handler(connection, message)
  File "/usr/lib/python2.7/site-packages/easyirc/event/__init__.py", line 109, in __call__
    return self.job(connection, message)
  File "/usr/lib/python2.7/site-packages/easyirc/event/__init__.py", line 144, in job
    return self.job_func(connection, msgline.sender, *util.msgsplit(message)[2:])
  File "IEChatBot.py", line 237, in on_privmsg
    botreply = self.rs.reply(nick, message)
  File "/usr/lib/python2.7/site-packages/rivescript/rivescript.py", line 1469, in reply
    reply = self._getreply(user, msg)
  File "/usr/lib/python2.7/site-packages/rivescript/rivescript.py", line 1651, in _getreply
    match = re.match(r'^' + regexp + r'$', msg)
  File "/usr/lib64/python2.7/re.py", line 141, in match
    return _compile(pattern, flags).match(string)
  File "/usr/lib64/python2.7/re.py", line 251, in _compile
    raise error, v # invalid expression
```

By inserting a print statement in rivescript/rivescript.py line 1650 I was able to see the buggy generated regex:

`tell (me|us) more(?:\s*(about (it\s*|\s*ie\s*|\s*)\s*|\s*about [the\s*|\s*)intel exchange)] `

As you can see, the problem is just after the optional "the": Insted of processing the parantheses, they were skipped by the code generating the regexes, which I traced down to rivescript/rivescript.py lines 1858-1869. 

A minimal example for replicating the issue:

In a python file
```
import rivescript

rs = rivescript.RiveScript()
rs.load_file("bugreport.rive")
rs.sort_replies()
rs.reply("test", "anything_here")
```

In bugreport.rive:
```
! version = 2.0

+ tell (me|us) more [(about (it|ie|)|about [the] intel exchange)]
- proof
```

**Comment by kirsle** [comment]
Nesting alternations and optionals like this isn't a supported feature. Triggers don't call themselves regular expressions, so don't think of them as regular expressions.

The trigger format is meant to be simplistic so that the user writing a chatbot doesn't have to understand regexps to write them (and if you tried to anyway, you wouldn't get very far because some regexp metacharacters such as `*`, `[]`, `()` and `{}` have different special meanings to RiveScript). It would also confuse users with the capture-groups on their triggers. For example,

`tell (me|us) more [(about (it|ie|)|about [the] intel exchange)]`

* What's in `<star1>`? Probably `me|us`, that part's easy.
* What about `<star2>`? It would probably be the entire thing inside of the outer set of square brackets. But that outer set of brackets is a non-capturing group (optionals never go into `<star>` tags), so if the group is missing what would `<star2>` be? I don't know. It could be `None` or `"undefined"` or raise an exception.
* `<star3>` is in an even more indeterminate state than this. `<star2>` might *encompass* `<star3>` but if the user matched the other half of the second capture group, who knows what `<star3>` would contain then.

So basically... don't do that. :wink: 

I should probably update official documentation saying that this is NOT a supported feature. But for example, in [aiml2rs](https://github.com/aichaos/aiml2rs), it is listed under one of the caveats when converting RiveScript into AIML.

> Nested parenthesis groups in triggers will be skipped. These are triggers where you include optionals INSIDE an alternation group, or any similar combination. These kinds of triggers can't be permutated cleanly, so are skipped.
